src/codegen/doc/format.rs

```diff
@@ -0,0 +1,73 @@
+use regex::{Captures, Regex};
+
+const LANGUAGE_SEP_END : &'static str = "\" -->";
+const LANGUAGE_BLOCK_BEGIN : &'static str = "|[<!-- language=\"";
+const LANGUAGE_BLOCK_END : &'static str = "\n]|";
+
+lazy_static! {
+    static ref REG : Regex = Regex::new(r"#?(?:G[A-Z][a-z]|G[dt]k)[\w]*").unwrap();
+    static ref REG2 : Regex = Regex::new(r"@\w*(\s|\n)").unwrap();
+}
+
+pub fn reformat_doc(input: &str) -> String {
+    code_blocks_transformation(input)
+}
+
+fn code_blocks_transformation(input: &str) -> String {
+    let mut out = String::new();
+
+    for (num, entry) in input.split(LANGUAGE_BLOCK_BEGIN).into_iter().enumerate() {
```

Comment:
If you insist in doing this the hard way, I think the primitive you need here is a custom `split` like this:
```
fn try_split<'a>(src: &'a str, needle: &src) -> (&'a str, Option<&'a str>) {
    match src.find(needle) {
        Some(pos) => (src[..pos], Some(src[pos + needle.len())),
        None => (src, None),
    }
}
```
Then you'll only need one loop, likely without any `break`/`continue`.

Regex is going to be more robust still and easily handle optional language and make sure `|[` starts at newline if you ask it to.